File: internal/extract/extract.go
// Package extract wraps ffmpeg to pull frames and audio from a video file.
// Frames are extracted at 1 FPS (configurable), scaled to 640px wide.
// Audio is extracted as 16kHz mono WAV — Gemma 4's native input rate.
package extract

import (
	"encoding/base64"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
)

// Frame holds a single video frame with its timestamp.
type Frame struct {
	TimestampS float64
	Base64JPEG string
	Path       string
}

// ExtractFrames pulls frames from video at the given FPS into a temp dir.
// Returns at most maxFrames frames starting from startS.
func ExtractFrames(videoPath string, startS, durationS float64, fps, maxFrames int) ([]Frame, error) {
	dir, err := os.MkdirTemp("", "tsi_frames_*")
	if err != nil {
		return nil, fmt.Errorf("create temp dir: %w", err)
	}

	args := []string{
		"-ss", fmt.Sprintf("%.3f", startS),
		"-i", videoPath,
		"-t", fmt.Sprintf("%.3f", durationS),
		"-vf", fmt.Sprintf("fps=%d,scale=640:-1", fps),
		"-q:v", "3",
		filepath.Join(dir, "frame_%04d.jpg"),
		"-y", "-loglevel", "error",
	}

	if out, err := exec.Command("ffmpeg", args...).CombinedOutput(); err != nil {
		return nil, fmt.Errorf("ffmpeg frames: %w\n%s", err, out)
	}

	entries, err := filepath.Glob(filepath.Join(dir, "frame_*.jpg"))
	if err != nil {
		return nil, err
	}
	sort.Strings(entries)

	if maxFrames > 0 && len(entries) > maxFrames {
		entries = entries[:maxFrames]
	}

	frames := make([]Frame, 0, len(entries))
	for i, path := range entries {
		data, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		frames = append(frames, Frame{
			TimestampS: startS + float64(i)/float64(fps),
			Base64JPEG: base64.StdEncoding.EncodeToString(data),
			Path:       path,
		})
	}
	return frames, nil
}

// ExtractAudio pulls a WAV segment (16kHz mono) from the video.
// Returns the path to the temp WAV file.
func ExtractAudio(videoPath string, startS, durationS float64) (string, error) {
	f, err := os.CreateTemp("", "tsi_audio_*.wav")
	if err != nil {
		return "", fmt.Errorf("create temp wav: %w", err)
	}
	f.Close()

	args := []string{
		"-ss", fmt.Sprintf("%.3f", startS),
		"-i", videoPath,
		"-t", fmt.Sprintf("%.3f", durationS),
		"-ar", "16000", // 16kHz — Gemma 4 native rate
		"-ac", "1",     // mono
		"-f", "wav",
		f.Name(),
		"-y", "-loglevel", "error",
	}

	if out, err := exec.Command("ffmpeg", args...).CombinedOutput(); err != nil {
		return "", fmt.Errorf("ffmpeg audio: %w\n%s", err, out)
	}
	return f.Name(), nil
}

// AudioToBase64 reads a WAV file and returns its base64 encoding.
func AudioToBase64(wavPath string) (string, error) {
	data, err := os.ReadFile(wavPath)
	if err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(data), nil
}

// Duration returns the video duration in seconds via ffprobe.
func Duration(videoPath string) (float64, error) {
	out, err := exec.Command(
		"ffprobe", "-v", "error",
		"-show_entries", "format=duration",
		"-of", "default=noprint_wrappers=1:nokey=1",
		videoPath,
	).Output()
	if err != nil {
		return 0, fmt.Errorf("ffprobe: %w", err)
	}
	s := strings.TrimSpace(string(out))
	d, err := strconv.ParseFloat(s, 64)
	if err != nil {
		return 0, fmt.Errorf("parse duration %q: %w", s, err)
	}
	return d, nil
}

// Cleanup removes temp files created during extraction.
func Cleanup(paths ...string) {
	for _, p := range paths {
		os.Remove(p)
		// Also try removing parent dir if it looks like our temp dir
		dir := filepath.Dir(p)
		if strings.Contains(filepath.Base(dir), "tsi_") {
			os.RemoveAll(dir)
		}
	}
}

File: extract.py
"""
Video extraction utilities.
Pulls frames at 1 FPS and raw audio at 16kHz — matching Gemma 4 12B's native input format.
"""

import subprocess
import tempfile
import os
from pathlib import Path
from PIL import Image
import base64
import io


def extract_frames(video_path: str, fps: int = 1, max_frames: int = 60) -> list[dict]:
    """
    Extract frames from video at given FPS.
    Returns list of {timestamp_s, base64_jpeg} dicts.
    """
    out_dir = tempfile.mkdtemp(prefix="tsi_frames_")
    cmd = [
        "ffmpeg", "-i", video_path,
        "-vf", f"fps={fps},scale=640:-1",
        "-q:v", "3",
        f"{out_dir}/frame_%04d.jpg",
        "-y", "-loglevel", "error"
    ]
    subprocess.run(cmd, check=True)

    frames = []
    frame_files = sorted(Path(out_dir).glob("frame_*.jpg"))[:max_frames]

    for i, f in enumerate(frame_files):
        timestamp = i / fps
        img = Image.open(f)
        buf = io.BytesIO()
        img.save(buf, format="JPEG", quality=85)
        b64 = base64.b64encode(buf.getvalue()).decode()
        frames.append({"timestamp_s": timestamp, "b64": b64, "path": str(f)})

    return frames


def extract_audio_segment(video_path: str, start_s: float, duration_s: float = 30.0) -> str:
    """
    Extract a raw audio segment as a temp WAV file (16kHz mono).
    Gemma 4 12B ingests 16kHz audio projected directly into token space.
    """
    out_path = tempfile.mktemp(suffix=".wav", prefix="tsi_audio_")
    cmd = [
        "ffmpeg", "-i", video_path,
        "-ss", str(start_s),
        "-t", str(duration_s),
        "-ar", "16000",   # 16kHz — Gemma 4 12B's native rate
        "-ac", "1",       # mono
        "-f", "wav",
        out_path,
        "-y", "-loglevel", "error"
    ]
    subprocess.run(cmd, check=True)
    return out_path


def get_video_duration(video_path: str) -> float:
    """Return video duration in seconds via ffprobe."""
    result = subprocess.run(
        ["ffprobe", "-v", "error", "-show_entries", "format=duration",
         "-of", "default=noprint_wrappers=1:nokey=1", video_path],
        capture_output=True, text=True, check=True
    )
    return float(result.stdout.strip())


def audio_to_base64(wav_path: str) -> str:
    with open(wav_path, "rb") as f:
        return base64.b64encode(f.read()).decode()
